packages/nimble-components/src/rich-text/editor/index.ts

```diff
@@ -335,6 +335,35 @@ export class RichTextEditor extends FoundationElement implements ErrorPattern {
 
     private createTiptapEditor(): Editor {
         const customLink = this.getCustomLinkExtension();
+        const validLinkRegex = /^https?:\/\//i;
+
+        /**
+         * @param htmlString contains the html string of the copied content. If the content is a link, the `htmlString` contains anchor tag and a href value.
+         * ProseMirror reference for `transformPastedHTML`: https://prosemirror.net/docs/ref/#view.EditorProps.transformPastedHTML
+         */
+        const transformPastedHTML = function (htmlString: string): string {
+            const tempElement = document.createElement('div');
+            tempElement.innerHTML = htmlString;
```

Comment:
On second thought, is it handled internally and rendering it just as plain text as we switched off HTML? I know this happens while parsing markdown but just wanted to make sure that it also happens after this function hands over the string back to Prosemirror/Tiptap?

Reply:
Since the `transformPastedHTML` accepts and returns HTML in the form of a string, it is already ignoring tags like `script` and I think they are handled within the ProseMirror. Here is the snapshot when pasting a `<script>alert('attack')</script>` string and trying to extract the input in this method,
![image](https://github.com/ni/nimble/assets/123377523/7f28ac39-d15e-4378-a92d-8ac2700a1206)

As in the image, the tag values are escaped for the angular brackets of script tags and paste only the text content of it.

Also, I tried to modify the code by creating a script element just like the above example and returning its `innerHTML` it doesn't cause any issues. Since the HTML string is passed through this method every time and we just have it in the HTML element returning the same `innerHTML` and modifying only for anchor elements, it seems to be working as expected. I also wrote a test case to add script tags in the anchor tags and verified it renders as plain text.

Reply:
After searching around for a bit I think relying on the template element and it's document context should give an inert context for the HTML manipulation:


I'm having trouble finding documentation that manipulating the dom nodes of a template in a document fragment in this way actually preserves the safety guarantee of the template element..

```ts
const validAbsoluteLinkRegex = /^https?:\/\//i;
document.createElement('template');
template.innerHTML = '<a href="HTTPS://NIMBLE.NI.DEV">Nimble</a>';
const templateDocument = template.content.ownerDocument;
template.content.querySelectorAll('a').forEach(anchorElement => {
        const href = anchorElement.getAttribute('href');
        // When pasting a link, the `href` attribute of the anchor element should be a valid HTTPS/HTTP link;
        // else, it should be rendered as plain text in a paragraph element.
        if (href && validAbsoluteLinkRegex.test(href)) {
            anchorElement.textContent = href; // Modifying the anchor element text content with its href
        } else {
            const paragraphElement = templateDocument.createElement('p');
            paragraphElement.textContent = anchorElement.textContent;

            // TODO under what circumstance will the parentNode be nullish?
            // Do you have a test case for this?
            anchorElement.parentNode?.replaceChild(
                paragraphElement,
                anchorElement
            );
        }
    });

console.log(template.innerHTML);
```

Reply:
Regarding point 3, we have restricted the image pasting event by utilizing the [`handlePaste`](https://prosemirror.net/docs/ref/#view.EditorProps.handlePaste) method and prevent the event when the image or any files are uploaded to the editor.

Reply:
As discussed offline, the current implementation is better than using the HTML strings so I am resolving this comment.